### DISCUSSION TITLE
Schema is not picked up when in a different file

### DISCUSSION THREAD

**Discussion by gino** [hayes/pothos#166] [Q&A] [answered]
Hey everyone,

I am trying to setup GiraphQL while having the following file structure:
- `src/builder.ts` that exports the `SchemaBuilder`
- `src/index.ts` that creates, in my case, a GraphiQL server and passes the `builder.toSchema({})` to that server.

And then I have a `src/schema/user.ts` with a basic query copied from the documentation:

```ts
import builder from "../builder";

builder.queryType({
  fields: (t) => ({
    hello: t.string({
      args: {
        name: t.arg.string(),
      },
      resolve: (parent, { name }) => `hello, ${name || "World"}`,
    }),
  }),
});
```

However, I expected that the builder would pick this query up while it is in a different file. But unfortunately, the actual schema stays empty and GraphiQL obviously gives me the following error:
```json
{
  "errors": [
    {
      "message": "Query root type must be provided."
    }
  ]
}
```

Which makes sense, since the schema is empty.

Is there another way to make sure that I could create queries and mutations in different files and make sure that those will get picked up by the builder and in a schema. I am just trying GiraphQL out and I might be doing something wrong, I think that this should work though.

Let me know if I am missing something. Thanks a lot!

**Comment by hayes** [comment, marked as answer]
You just need to import the files that define your types before you call .toSchema.  I usually have a schema.ts file that imports all my files that define types, then have that file export the built schema

**Reply by gino** [reply]
Hmm that makes sense. I got that working by just doing this manually, for now:
```ts
require("./schema/test");
require("./schema/user");

const schema = builder.toSchema({});

export default schema;
```

However, I have a `queryType` defined in both files and unfortunately they conflict: `Duplicate typename: Another type with name Query already exists`. I am not exporting anything from the actual schema types (user.ts & test.ts) so I just require them like above. I shouldn't need to export anything I think.

However, I could try to merge the types with some other tools, not sure if that's the best approach. Pretty sure others have tried to achieve the same, just not too sure how, without exporting anything from the schema types.

I also looked into extending the type and found an old [plugin](https://www.npmjs.com/package/@giraphql/plugin-extends) for GiraphQL, unfortunately not working with the latest release though.

**Reply by hayes** [reply]
You can't define a type multiple times because there are many things that can't be easily merged (description, auth settings, etc).  What you are probably looking for are methods like queryField, queryFields, object field, etc.  These let you define one or multiple fields for a type without actually re-defining the type.  These methods are what replaced the extends plugin

**Reply by gino** [reply]
Ah I see now. Yeah I was a bit confused by that, thanks a lot. That's great!
